clippy_lints/src/assigning_clones.rs

```diff
@@ -0,0 +1,322 @@
+
+    // Fast paths to only check method calls without arguments or function calls with a single argument
+    let (target, kind, resolved_method) = match expr.kind {
+        ExprKind::MethodCall(path, receiver, [], _span) => {
+            let args = cx.typeck_results().node_args(expr.hir_id);
+
+            // If we could not resolve the method, don't apply the lint
+            let Ok(Some(resolved_method)) = Instance::resolve(cx.tcx, cx.param_env, fn_def_id, args) else {
+                return None;
+            };
+            if is_trait_method(cx, expr, sym::Clone) && path.ident.name == sym::clone {
+                (TargetTrait::Clone, CallKind::MethodCall { receiver }, resolved_method)
+            } else if is_trait_method(cx, expr, sym::ToOwned) && path.ident.name.as_str() == "to_owned" {
+                (TargetTrait::ToOwned, CallKind::MethodCall { receiver }, resolved_method)
+            } else {
+                return None;
+            }
+        },
+        ExprKind::Call(function, [arg]) => {
+            let kind = cx.typeck_results().node_type(function.hir_id).kind();
+
+            // If we could not resolve the method, don't apply the lint
+            let Ok(Some(resolved_method)) = (match kind {
+                ty::FnDef(_, args) => Instance::resolve(cx.tcx, cx.param_env, fn_def_id, args),
+                _ => Ok(None),
+            }) else {
+                return None;
+            };
+            if cx.tcx.is_diagnostic_item(sym::to_owned_method, fn_def_id) {
+                (
+                    TargetTrait::ToOwned,
+                    CallKind::FunctionCall { self_arg: arg },
+                    resolved_method,
+                )
+            } else if let Some(trait_did) = cx.tcx.trait_of_item(fn_def_id)
+                && cx.tcx.is_diagnostic_item(sym::Clone, trait_did)
+            {
+                (
+                    TargetTrait::Clone,
+                    CallKind::FunctionCall { self_arg: arg },
+                    resolved_method,
+                )
+            } else {
+                return None;
+            }
+        },
+        _ => return None,
+    };
+
+    Some(CallCandidate {
+        target,
+        kind,
+        method_def_id: resolved_method.def_id(),
+    })
+}
+
+// Return true if we find that the called method has a custom implementation and isn't derived or
+// provided by default by the corresponding trait.
+fn is_ok_to_suggest<'tcx>(cx: &LateContext<'tcx>, lhs: &Expr<'tcx>, call: &CallCandidate<'tcx>) -> bool {
+    // If the left-hand side is a local variable, it might be uninitialized at this point.
+    // In that case we do not want to suggest the lint.
+    if let Some(local) = path_to_local(lhs) {
+        // TODO: This check currently bails if the local variable has no initializer.
+        // That is overly conservative - the lint should fire even if there was no initializer,
+        // but the variable has been initialized before `lhs` was evaluated.
```

Comment:
Here's the HIR representation for that:
```
a = 4.clone();
```

```
[clippy_lints/src/check_exprs.rs:28:9] expr = Expr {
    hir_id: HirId(DefId(0:3 ~ check_exprs[0844]::main).10),
    kind: Assign(
        Expr {
            hir_id: HirId(DefId(0:3 ~ check_exprs[0844]::main).11),
            kind: Path(
                Resolved(
                    None,
                    Path {
                        span: tests/ui/check_exprs.rs:6:5: 6:6 (#0),
                        res: Local(
                            HirId(DefId(0:3 ~ check_exprs[0844]::main).4),
                        ),
                        segments: [
                            PathSegment {
                                ident: a#0,
                                hir_id: HirId(DefId(0:3 ~ check_exprs[0844]::main).12),
                                res: Local(
                                    HirId(DefId(0:3 ~ check_exprs[0844]::main).4),
                                ),
                                args: None,
                                infer_args: true,
                            },
                        ],
                    },
                ),
            ),
            span: tests/ui/check_exprs.rs:6:5: 6:6 (#0),
        },
        Expr {
            hir_id: HirId(DefId(0:3 ~ check_exprs[0844]::main).13),
            kind: MethodCall(
                PathSegment {
                    ident: clone#0,
                    hir_id: HirId(DefId(0:3 ~ check_exprs[0844]::main).14),
                    res: Err,
                    args: None,
                    infer_args: true,
                },
                Expr {
                    hir_id: HirId(DefId(0:3 ~ check_exprs[0844]::main).15),
                    kind: Lit(
                        Spanned {
                            node: Int(
                                Pu128(
                                    4,
                                ),
                                Unsuffixed,
                            ),
                            span: tests/ui/check_exprs.rs:6:9: 6:10 (#0),
                        },
                    ),
                    span: tests/ui/check_exprs.rs:6:9: 6:10 (#0),
                },
                [],
                tests/ui/check_exprs.rs:6:11: 6:18 (#0),
            ),
            span: tests/ui/check_exprs.rs:6:9: 6:18 (#0),
        },
        tests/ui/check_exprs.rs:6:7: 6:8 (#0),
    ),
    span: tests/ui/check_exprs.rs:6:5: 6:18 (#0),
}
```

The pattern matching should be simple :) If you're out of ideas, I can also give it a go and commit it directly.

Reply:
I'm not sure how this relates to pattern matching on HIR :thinking: To fully resolve this, we would need dataflow analysis, right?

E.g.
```rust
let mut x: u32;
if foo() {
   x = 5;
   x = y.clone(); // should fire
} else {
   x = y.clone(); // should not fire
}
```